io-engine/src/lvs/lvs_lvol.rs

```diff
@@ -443,15 +442,108 @@ impl Lvol {
 
         // Wait till operation succeeds, if requested.
         match receiver {
-            None => Ok(()),
-            Some(r) => r
-                .await
-                .expect("Snapshot done callback disappeared")
-                .to_result(|error| Error::SnapshotCreate {
-                    source: Errno::from_i32(error),
-                    msg: c_snapshot_name.into_string().unwrap(),
-                }),
+            None => Ok(None),
+            Some(receiver) => {
+                let (error, lvol) =
+                    receiver.await.expect("Snapshot done callback disappeared");
+                match error {
+                    0 => Ok(Some(lvol)),
+                    _ => Err(Error::SnapshotCreate {
+                        source: Errno::from_i32(error),
+                        msg: c_snapshot_name.into_string().unwrap(),
+                    }),
+                }
+            }
+        }
+    }
+    /// List All Snapshot.
+    pub fn list_all_snapshots() -> Vec<VolumeSnapshotDescriptor> {
+        let mut snapshot_list: Vec<VolumeSnapshotDescriptor> = Vec::new();
+        let bdev =
+            UntypedBdev::bdev_first().expect("Failed to enumerate devices");
+
+        let lvol_devices = bdev
+            .into_iter()
+            .filter(|b| b.driver() == "lvol")
+            .map(|b| Lvol::try_from(b).unwrap())
+            .collect::<Vec<Lvol>>();
+        // return empty snapshot parameter list, if no snapshot found.
+        if lvol_devices.len() <= 1 {
+            return snapshot_list;
+        }
+        for lvol in lvol_devices {
+            // skip lvol if it is not snapshot.
+            if !lvol.is_snapshot() {
+                continue;
+            }
+            let blob = lvol.bs_iter_first();
+            let mut skip_lvol = false;
+            let mut snapshot_param: SnapshotParams = Default::default();
+            for attr in SnapshotXattrs::iter() {
+                let mut val = std::ptr::null();
+                let mut size: u64 = 0;
+                let attr_id = attr.name().to_string().into_cstring();
+                let curr_attr_val;
+                unsafe {
```

Comment:
this code seems a bit repeated, can we have a helper fn which can be called from multiple locations?

Reply:
Added an utility function.